src/config.rs

```diff
@@ -0,0 +1,248 @@
+            }
+            PackageConfig::License { license } => package_record
+                .license
+                .is_some_and(|l| l.as_str() == license),
+        }
+    }
+}
+
+#[derive(Deserialize, Debug, Clone)]
+#[serde(deny_unknown_fields, rename_all = "kebab-case")]
+pub struct S3Config {
+    pub endpoint_url: Url,
+    pub region: String,
+    pub force_path_style: bool,
+}
+
+// TODO: allow setting it in .s3-config globally for both source and dest
+#[derive(Deserialize, Debug, Clone)]
+#[serde(deny_unknown_fields, rename_all = "kebab-case")]
+pub struct S3ConfigSourceDest {
+    pub source: Option<S3Config>,
+    pub destination: Option<S3Config>,
+}
+
+#[derive(Deserialize, Debug, Clone, Default)]
+#[serde(deny_unknown_fields, rename_all = "kebab-case")]
+pub struct CondaMirrorYamlConfig {
+    pub source: Option<NamedChannelOrUrl>,
+    pub destination: Option<NamedChannelOrUrl>,
+    pub subdirs: Option<Vec<Platform>>,
+
+    pub include: Option<Vec<PackageConfig>>,
+    pub exclude: Option<Vec<PackageConfig>>,
+    pub s3_config: Option<S3ConfigSourceDest>,
+}
+
+/* -------------------------------------------- CONFIG ------------------------------------------- */
+
+#[derive(Debug, Clone)]
+pub enum MirrorMode {
+    /// Mirror all packages.
+    All,
+    /// Mirror all packages except those matching the given patterns.
+    AllButExclude(Vec<PackageConfig>),
+    /// Mirror only packages matching the given patterns.
+    OnlyInclude(Vec<PackageConfig>),
+    /// Mirror all packages except those matching the given patterns.
+    /// Override excludes with include patterns.
+    IncludeExclude(Vec<PackageConfig>, Vec<PackageConfig>),
+}
+
+#[derive(Clone)]
+pub struct CondaMirrorConfig {
+    pub source: NamedChannelOrUrl,
+    pub destination: NamedChannelOrUrl,
+    pub subdirs: Option<Vec<Platform>>,
+    pub mode: MirrorMode,
+    pub s3_config_source: Option<S3Config>,
+    pub s3_config_destination: Option<S3Config>,
+    // pub s3_credentials_source: Option<S3Credentials>,
+    pub s3_credentials_destination: Option<S3Credentials>,
+}
+
+impl std::fmt::Debug for CondaMirrorConfig {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("CondaMirrorConfig")
+            .field("source", &self.source)
+            .field("destination", &self.destination)
+            .field("subdirs", &self.subdirs)
+            .field("mode", &self.mode)
+            .field("s3_config_source", &self.s3_config_source)
+            .field("s3_config_destination", &self.s3_config_destination)
+            // .field("s3_credentials_source", &{
+            //     if self.s3_credentials_source.is_some() {
+            //         "Some(***)"
+            //     } else {
+            //         "None"
+            //     }
+            // })
+            .field("s3_credentials_destination", &{
+                if self.s3_credentials_destination.is_some() {
+                    "Some(***)"
```

Comment:
If you wanted to, you could also create a "newtype" for the string, and a custom impl for ToString or Debug, and then you could skip the rest of this code.

E.g. 

```
struct S3Credential(String);

impl FromStr for S3Credential ...

impl Debug ... 
```

Then no-one else could make this mistake of printing the value.

Reply:
you mean in rattler?